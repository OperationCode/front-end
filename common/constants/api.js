import { get, post, patch } from 'common/utils/api-utils';
import { formatUserData } from 'common/utils/formatters';

/* GET REQUESTS */
export const getUserPromise = ({ token }) => get('auth/user/', { token });
export const getCodeSchoolsPromise = () => get('api/v1/codeschools/');
export const getTeamMembersPromise = () => get('api/v1/teamMembers/');

/* POST REQUESTS */
export const createUser = ({ email, password, firstName, lastName, zipcode }) =>
  post('auth/registration/', {
    email,
    firstName,
    lastName,
    password,
    zipcode,
  }).then(({ data }) => data);

export const loginUser = ({ email, password }) =>
  post('auth/login/', {
    email,
    password,
  }).then(({ data }) => data);

export const loginSocial = (provider, body) =>
  post(`auth/social/${provider}/`, body).then(({ data }) => data);

export const passwordReset = ({ email }) =>
  post('auth/password/reset/', { email }).then(({ data }) => data);

export const passwordResetSubmit = values =>
  post('auth/password/reset/confirm/', values).then(({ data }) => data);

export const changePassword = values =>
  post('auth/password/change/', values).then(({ data }) => data);

export const confirmEmail = key => post('auth/verify-email/', key).then(({ data }) => data);

/* PATCH REQUESTS */
export const updateUser = userInfo => {
  return patch('auth/profile/', {
    ...formatUserData(userInfo),
  }).then(({ data }) => data);
};
