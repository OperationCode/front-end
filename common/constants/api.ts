import { patch, post } from 'common/utils/api-utils';
import type { RegistrationFormValues } from 'components/Forms/RegistrationForm/RegistrationForm';
import type { MilitaryStatusFormShape } from 'components/Forms/UpdateProfileForm/steps/MilitaryStatus';
import type { ProfessionalDetailsFormShape } from 'components/Forms/UpdateProfileForm/steps/ProfessionalDetails';

export const registerUser = (values: RegistrationFormValues) =>
  post('/api/registration/new', values);

export const updateUser = (values: ProfessionalDetailsFormShape | MilitaryStatusFormShape) =>
  patch('/api/registration/update', values);
