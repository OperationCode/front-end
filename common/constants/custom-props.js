import PropTypes from 'prop-types';

/*
 * When declaring a new, commonly utilized prop type, please also export an object that can be used
 * by the Knobs Storybook addon easily. See Button's default story with analyticsObject for example.
 * NOTE: Export it as a function so that the object isn't mutable across multiple stories.
 */

/* eslint-disable import/prefer-default-export */
// See: https://github.com/react-ga/react-ga#reactgaeventargs for reference
export const googleAnalyticsEventPropType = PropTypes.shape({
  // A description of the behaviour. E.g. 'Clicked Delete', 'Added a component', 'Deleted account'
  action: PropTypes.string.isRequired,

  // A top level category for these events. E.g. 'User', 'Navigation', 'App Editing', etc.
  category: PropTypes.string.isRequired,

  // More precise labelling of the related action. E.g. alongside the 'Added a component' action,
  // we could add the name of a component as the label. E.g. 'Survey', 'Heading', 'Button', etc.
  label: PropTypes.string,

  // A means of recording a numerical value against an event. E.g. a rating, a score, etc.
  value: PropTypes.number,

  // If an event is not triggered by a user interaction, but instead by our code (e.g. on page
  // load, it should be flagged as a nonInteraction event to avoid skewing bounce rate data.
  nonInteraction: PropTypes.boolean,

  // This specifies the transport mechanism with which hits will be sent. Valid values include
  // 'beacon', 'xhr', or 'image'.
  transport: PropTypes.oneOf(['beacon', 'xhr', 'image']),
});

export const googleAnalyticsEventStoryObjectFactory = () => ({
  action: '',
  category: '',
  label: '',
  value: undefined,
  nonInteraction: false,
  transport: undefined,
});
