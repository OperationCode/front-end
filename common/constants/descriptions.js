/* eslint-disable max-len */
/** Provides lorem ipsum descriptions of length
 * short: 1 sentence,
 * medium: 3 sentences,
 * long: 5 sentences
 */
export const descriptions = {
  short: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',

  medium:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut in semper justo, quis efficitur urna. Aliquam et pretium est, nec viverra risus. Curabitur facilisis est in nibh luctus condimentum.',

  long:
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Ut in semper justo, quis efficitur urna. Aliquam et pretium est, nec viverra risus. Curabitur facilisis est in nibh luctus condimentum. Maecenas sed urna sed est dapibus venenatis. Quisque aliquet velit eu ligula pellentesque, non pretium massa blandit. Suspendisse vitae viverra ante, eget sodales arcu. Aliquam erat volutpat.',
};
