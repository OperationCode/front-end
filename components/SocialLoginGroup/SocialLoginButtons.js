import React from 'react';
import { func } from 'prop-types';
import classNames from 'classnames';
import GoogleLogin from 'react-google-login';
import FacebookLogin from 'react-facebook-login/dist/facebook-login-render-props';
import { clientTokens } from 'common/config/environment';
import styles from './SocialLoginGroup.css';

SocialLoginButtons.propTypes = {
  onSuccess: func.isRequired,
  onGoogleFailure: func.isRequired,
};

function SocialLoginButtons({ onSuccess, onGoogleFailure }) {
  return (
    <>
      <GoogleLogin
        clientId={clientTokens.OC_GOOGLE_KEY}
        buttonText="Login"
        onSuccess={onSuccess('google')}
        onFailure={onGoogleFailure}
        render={renderProps => (
          <button
            type="button"
            onClick={renderProps.onClick}
            disabled={renderProps.disabled}
            className={classNames(styles.loginButton, styles.googleButton)}
          >
            Login with Google
          </button>
        )}
      />

      <FacebookLogin
        appId={clientTokens.OC_FACEBOOK_KEY}
        callback={onSuccess('facebook')}
        redirectUri={typeof window === 'object' && `${window.location.origin}/login`}
        render={renderProps => (
          <button
            type="button"
            onClick={renderProps.onClick}
            className={classNames(styles.loginButton, styles.facebookButton)}
          >
            Login with Facebook
          </button>
        )}
      />
    </>
  );
}

export default SocialLoginButtons;
