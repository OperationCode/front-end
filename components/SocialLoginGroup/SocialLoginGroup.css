.SocialLoginGroup {
  width: 100%;
}

.flexRow {
  display: flex;
  flex-flow: row wrap;
  justify-content: center;
  align-items: center;
}

.fullWidth {
  width: 100%;
}

/* Shared */
.loginButton {
  font-family: Helvetica, sans-serif;
  width: 260px;
  box-sizing: border-box;
  position: relative;
  margin: 0.2em;
  padding: 0 15px 0 46px;
  border: none;
  text-align: left;
  line-height: 34px;
  white-space: nowrap;
  border-radius: 0.2em;
  font-size: 14px;
  color: #ffffff;
}

.loginButton::before {
  content: "";
  box-sizing: border-box;
  position: absolute;
  top: 0;
  left: 0;
  width: 34px;
  height: 100%;
}

.loginButton:active {
  opacity: 0.75;
  box-shadow: inset 0 0 0 32px rgba(0, 0, 0, 0.1);
}

/* Alert Message */
.alertContainer {
  margin-bottom: 1rem;
}

.alertFill {
  max-width: 500px;
  margin: 0 auto;
}

/* Facebook */
.facebookButton {
  background-color: #4c69ba;
  background-image: linear-gradient(#4c69ba, #3b55a0);
  text-shadow: 0 -1px 0 #354c8c;
  transition: opacity 0.15s linear;
}

.facebookButton::before {
  border-right: #364e92 1px solid;
  background: url("https://s3-us-west-2.amazonaws.com/s.cdpn.io/14082/icon_facebook.png") 6px 6px
    no-repeat;
}

.facebookButton:active,
.facebookButton:hover,
.facebookButton:focus {
  opacity: 0.8;
  cursor: pointer;
}

/* Google */
.googleButton {
  background: #dd4b39;
  transition: opacity 0.15s linear;
}

.googleButton::before {
  border-right: #bb3f30 1px solid;
  background: url("https://s3-us-west-2.amazonaws.com/s.cdpn.io/14082/icon_google.png") 6px 6px
    no-repeat;
}

.googleButton:active,
.googleButton:hover,
.googleButton:focus {
  opacity: 0.8;
  cursor: pointer;
}

/* Github */
.githubButton {
  color: #eff3f6;
  background: #24292e;
  transition: opacity 0.15s linear;
  border: 1px solid #1b1f23;
}

.githubButton::before {
  border-right: #b0b0b0 1px solid;
  background: url("/static/images/icons/github_octocat_logo.svg") 4px 4px no-repeat;
}

.githubButton:active,
.githubButton:hover,
.githubButton:focus {
  opacity: 0.8;
  cursor: pointer;
}
