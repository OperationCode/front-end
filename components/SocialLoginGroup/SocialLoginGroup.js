import React, { useState } from 'react';
import { string, func } from 'prop-types';
import classNames from 'classnames';
import { getServerErrorMessage } from 'common/utils/api-utils';
import Alert from 'components/Alert/Alert';
import styles from './SocialLoginGroup.module.css';

SocialLoginGroup.propTypes = {
  className: string,
  handleSuccess: func.isRequired,
  children: func.isRequired,
  loginSocial: func.isRequired,
};

SocialLoginGroup.defaultProps = {
  className: undefined,
};

function SocialLoginGroup(props) {
  const [errorMessage, setErrorMessage] = useState('');

  const onSuccess = provider => async ({ accessToken }) => {
    try {
      const { handleSuccess, loginSocial } = props;
      const result = await loginSocial(provider, { accessToken });
      handleSuccess(result);
    } catch (error) {
      setErrorMessage(getServerErrorMessage(error));
    }
  };

  const onGoogleFailure = () => {
    setErrorMessage("Couldn't log in with Google");
  };

  const { className, children } = props;

  return (
    <div className={classNames(className, styles.flexRow, styles.SocialLoginGroup)}>
      <div className={classNames(styles.alertContainer, styles.fullWidth)}>
        {errorMessage && (
          <Alert className={classNames(styles.flexRow, styles.alertFill)} type="error">
            {errorMessage}
          </Alert>
        )}
      </div>

      {children({ onSuccess, onGoogleFailure })}
    </div>
  );
}

export default SocialLoginGroup;
