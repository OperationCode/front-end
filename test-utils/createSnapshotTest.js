/* eslint-env jest */
import renderer from 'react-test-renderer';

/**
 * Used to create fully-rendered snapshot test.
 * Useful for reusable component smoke tests.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
};
