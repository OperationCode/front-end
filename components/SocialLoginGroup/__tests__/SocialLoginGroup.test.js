import React from 'react';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import { render, act } from '@testing-library/react';
import { loginSocial } from 'common/constants/api';
import OperationCodeAPIMock from 'test-utils/mocks/apiMock';
import SocialLoginGroup from '../SocialLoginGroup';

describe('SocialLoginGroup', () => {
  const socialReturnToken = { accessToken: 'abc123' };

  function renderWithHelpers() {
    const handleSuccessSpy = jest.fn();

    let renderProps;

    act(() => {
      render(
        <SocialLoginGroup
          className="test-class"
          loginSocial={loginSocial}
          handleSuccess={handleSuccessSpy}
        >
          {({ onSuccess, onGoogleFailure }) => {
            renderProps = { onSuccess, onGoogleFailure };
          }}
        </SocialLoginGroup>,
      );
    });

    return { handleSuccessSpy, renderProps };
  }

  beforeEach(() => {
    OperationCodeAPIMock.reset();
  });

  it('should render with required props', () => {
    createSnapshotTest(
      <SocialLoginGroup loginSocial={jest.fn()} handleSuccess={jest.fn()}>
        {() => {}}
      </SocialLoginGroup>,
    );
  });

  it('calls callbacks when onSuccess is triggered against correct provider', async () => {
    const providerName = 'google';

    OperationCodeAPIMock.onPost(`auth/social/${providerName}/`, socialReturnToken).reply(200, {
      ...socialReturnToken,
    });

    const { handleSuccessSpy, renderProps } = renderWithHelpers();

    const onSuccess = renderProps.onSuccess(providerName);
    await onSuccess(socialReturnToken);

    expect(handleSuccessSpy).toHaveBeenCalledWith(socialReturnToken);
  });

  it('does NOT call handleSuccess when loginSocial fails', async () => {
    const providerName = 'facebook';

    const { handleSuccessSpy, renderProps } = renderWithHelpers();

    OperationCodeAPIMock.onPost(`auth/social/${providerName}/`, socialReturnToken).reply(400, {
      error: 'User is already registered with this e-mail address.',
    });

    await act(async () => {
      const onSuccess = renderProps.onSuccess(providerName);
      await onSuccess(socialReturnToken);
    });

    expect(handleSuccessSpy).not.toHaveBeenCalled();
    expect(document.querySelector('div[role="alert"]').textContent).toStrictEqual(
      'User is already registered with this e-mail address.',
    );
  });
});
