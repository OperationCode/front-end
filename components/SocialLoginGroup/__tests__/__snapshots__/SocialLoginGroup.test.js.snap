// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialLoginGroup should render with required props 1`] = `
<div
  className="flexRow SocialLoginGroup"
>
  <div
    className="alertContainer fullWidth"
  >
    
  </div>
</div>
`;
